gobblin-core/src/main/java/gobblin/source/extractor/extract/google/webmaster/GoogleWebmasterDataFetcherImpl.java

```diff
@@ -0,0 +1,291 @@
+          LOG.info(String.format("Total size of the data set for market-%s on %s is %d", country, date, startRow));
+          entireSize = startRow;
+        }
+
+        //update starting row and filters before starting
+        startRow = 0;
+        filters = new ArrayList<>();
+        filters.add(countryFilter);
+        for (String p : pagesRoots) {
+          filters.add(GoogleWebmasterFilter.pageFilter(GoogleWebmasterFilter.FilterOperator.NOTCONTAINS, p));
+        }
+        for (String p : pagesNoRoots) {
+          if (_siteProperty.equals(p)) {
+            //Must NOT add site property to the exclusion list, otherwise no pages will be returned.
+            continue;
+          }
+          filters.add(GoogleWebmasterFilter.pageFilter(GoogleWebmasterFilter.FilterOperator.NOTCONTAINS, p));
+        }
+        LOG.info(String.format(
+            "Starting another round of fetching for all pages. Currently there are %d pages w/o roots, and %d pages with roots",
+            pagesNoRoots.size(), pagesRoots.size()));
+      }
+    }
+
+    return new ImmutableTriple<>(pagesNoRoots, pagesRoots, entireSize);
+  }
+
+  /**
+   * This method gives you all pages given the page prefixes/keywords/root paths.
+   *
+   * The detailed steps are as follows:
+   *
+   * 1. For each page prefix, construct a filter group with two filters. One for country; the other for containing that page prefix.
+   * 2. Request for all pages with the filter group. If the count is 5000, go to step 3; otherwise, end of the process, we've already got all pages.
+   * 3. Expand current prefix by appending a char from the list [a-z,0-9,%,_,-,/].
+   * 4. Perform another request for each new prefix at last step. If the number of pages is 5000, repeat steps 3~4; otherwise, end of the process, we've already got all pages.
+   * 5. Union all pages as the return value
+   *
+   * @param date
+   * @param dimensions requested dimensions
+   * @param pagePrefixes a pagePrefix must end with "/"
+   * @return all pages given the page prefixes/keywords, or page root paths.
+   * @throws IOException
+   */
+  private Set<String> pagesWithRootPaths(String date, List<Dimension> dimensions, ApiDimensionFilter countryFilter,
+      Set<String> pagePrefixes) throws IOException {
+
+    for (String prefix : pagePrefixes) {
+      Preconditions.checkArgument(prefix.endsWith("/"), "Starting prefix must end with '/'");
+    }
+    Country countryString = countryFilterToEnum(countryFilter);
+
+    List<ApiDimensionFilter> filters = new LinkedList<>();
+    filters.add(countryFilter);
+    filters.add(null); //placeholder for page filter
+
+    Set<String> uniquePages = new HashSet<>();
+    Deque<String> toProcess = new LinkedList<>(pagePrefixes);
+    while (!toProcess.isEmpty()) {
+      String prefix = toProcess.remove();
+      LOG.info("Current page prefix is " + prefix);
+      filters.set(1, GoogleWebmasterFilter.pageFilter(GoogleWebmasterFilter.FilterOperator.CONTAINS, prefix));
```

Comment:
The overhead of creating a new filters array is not that high (compared to making the getPages() remote call). I think it is safer to create a new instance. That is necessary if we want to parallelize the getPages() calls in the future.